util/timer/src/timer.rs

```diff
@@ -0,0 +1,687 @@
+    }
+
+    pub fn new_timer<T>(&self, name: TimerName, handler: Arc<T>) -> TimerApi
+    where
+        T: 'static + TimeoutHandler + Send + Sync, {
+        let mut timers = self.timers.write();
+        match timers.entry(name) {
+            Entry::Occupied(_) => unreachable!("Timer name was already occupied: {}", name),
+            Entry::Vacant(entry) => {
+                entry.insert(handler);
+                TimerApi {
+                    timer_name: name,
+                    scheduler: Arc::downgrade(&self.scheduler),
+                }
+            }
+        }
+    }
+}
+
+pub struct TimerApi {
+    timer_name: TimerName,
+    scheduler: Weak<Scheduler>,
+}
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum ScheduleError {
+    TokenAlreadyScheduled,
+    TokenNotScheduled,
+    TimerLoopDropped,
+}
+
+impl TimerApi {
+    pub fn schedule_once(&self, after: Duration, timer_token: TimerToken) -> Result<(), ScheduleError> {
+        self.scheduler.upgrade().ok_or(ScheduleError::TimerLoopDropped)?.schedule(
+            TimerId(self.timer_name, timer_token),
+            after,
+            None,
+        )
+    }
+
+    pub fn schedule_repeat(&self, after: Duration, timer_token: TimerToken) -> Result<(), ScheduleError> {
+        self.scheduler.upgrade().ok_or(ScheduleError::TimerLoopDropped)?.schedule(
+            TimerId(self.timer_name, timer_token),
+            after,
+            Some(after),
+        )
+    }
+
+    pub fn cancel(&self, timer_token: TimerToken) -> Result<bool, ScheduleError> {
+        self.scheduler.upgrade().ok_or(ScheduleError::TimerLoopDropped)?.cancel(TimerId(self.timer_name, timer_token))
+    }
+}
+
+struct Scheduler {
+    inner: Mutex<SchedulerInner>,
+    condvar: Condvar,
+}
+
+impl Scheduler {
+    fn new() -> Scheduler {
+        Scheduler {
+            inner: Mutex::new(SchedulerInner::new()),
+            condvar: Condvar::new(),
+        }
+    }
+
+    fn schedule(&self, timer_id: TimerId, after: Duration, repeat: Option<Duration>) -> Result<(), ScheduleError> {
+        let mut scheduler = self.inner.lock();
+        scheduler.schedule(timer_id, after, repeat)?;
+        self.condvar.notify_one();
+        Ok(())
+    }
+
+    fn cancel(&self, timer_id: TimerId) -> Result<bool, ScheduleError> {
+        let mut scheduler = self.inner.lock();
+        let result = scheduler.cancel(timer_id)?;
+        self.condvar.notify_all();
+        Ok(result)
+    }
+
+    fn run(&self, worker_queue: &WorkerQueue) {
+        let mut scheduler = self.inner.lock();
+        while !scheduler.stop {
+            let wait_for = scheduler.handle_timeout(worker_queue);
+            match wait_for {
+                Some(timeout) => {
+                    self.condvar.wait_for(&mut scheduler, timeout);
+                }
+                None => self.condvar.wait(&mut scheduler),
+            }
+        }
+    }
+}
+
+
+// For any 'id' (TimerId) in 'self.states', any of following conditions should be false.
```

Comment:
Is there exactly one false? Or is it possible that more than one are false?

Reply:
I clarified conditions.